### ISSUE TITLE
Having sshd enabled by default

### ISSUE THREAD

**Issue by Altai-man** [void-linux/void-mklive#40]
Hi. Thanks for all the hard work of keeping the distro working.
Not sure where this ticket should go, please move if needed.

Currently, as far as I know, at least for a lxde-flavored live image, the sshd daemon is installed && up by default, which is true for an installed system too.

While I suspect some distros may have the same policy, I wonder what is the need behind of it? Possibly, installation over ssh?

I am opening the issue as it was pretty odd to me to find out that Void has it. Having a sshd sticking out into the network may be pretty dangerous, especially considering that Void doesn't promote itself as a server distro, but instead as a general-purpose OS, so the end-user may have no idea that sshd is up at all.

**Comment by SirMetathyst** [comment]
I find that strange because sshd is installed and disabled by default on my non-gui musl installs

**Comment by Altai-man** [comment]
To have more details, I just installed the system yesterday from https://a-hel-fi.m.voidlinux.org/live/current/ using `void-live-x86_64-20181111-lxde.iso` image, sshd was enabled by default. That's why I found it odd and reported.

**Comment by bobertlo** [comment]
I just did an install last night from the glibc base image and thought it was odd that sshd was installed and NOT enabled when it booted.

**Comment by atweiden** [comment]
@bobertlo

are we talking about the same thing? when the stock ISO image boots, it autoenables and runs sshd, along with other services. on insecure networks this means you are giving anyone on the LAN root on your machine. why would you want that as the default behavior?

**Comment by pullmoll** [comment]
I guess we could create two startup options for isolinux, one with sshd enabled and one disabled.
It should be easy to inspect `/proc/cmdline` for e.g. `void.sshd=no` in the `/etc/sv/sshd/run` script.

**Comment by the-maldridge** [comment]
@pullmoll that breaks the local install mechanism which copies the rootfs image to the target.  If we wanted to do something like that it would need to be a late start dracut module that checks for it.  Given that this is a core function of the images that makes it possible to remote install without a kvm, and that the solution to a questionably secure network is to just not join it until making the machine secure, I don't think this justifies engineering effort before the next ISO release.

Its also worth pointing out that there are a number of other things that should be done before just putting a machine on an untrusted network, most of which aren't candidates for inclusion in the images.

**Comment by Duncaen** [comment]
The dracut [script](https://github.com/void-linux/void-mklive/blob/master/dracut/vmklive/services.sh) which enables services on boot can check the kernel command line arguments.
It actually does this already with `live.services` to enable services that the script ignores.

**Comment by the-maldridge** [comment]
Given that Void does not market itself as a turn-key solution, and that there is a general assumption that Void users consider the security of their machines and networks they interact with, I'm going to close this issue.  If you cannot trust the network you are on during install, I highly recommend you locate a more secure network.

**Comment by noproto** [comment]
"Don't join an insecure network" is an insufficient response to exposing every user of Void Linux to be backdoored. Stating that it's a general assumption that the users are aware opaquely accepts the risk on behalf of _all users_ without minimum precautions such as informing them prior to installing (https://voidlinux.org/download/) that during the install process anybody can SSH to the system. I imagine quite a few users are unaware that during installation it has (all of) an SSH daemon listening on all interfaces with root login enabled, password authentication enabled, and default passwords set.

**Comment by Vaelatern** [comment]
Noproto, have you in fact verified that root login is enabled?

Also, if you think this is an inadequate answer, give a better one.

**Comment by noproto** [comment]
@Vaelatern I tested it prior to commenting. You can view it in the sshd_config:

> PermitRootLogin yes

RE: Providing a better answer, I suggested one in my previous comment (warning users prior to installation). Although, the best answer is to disable SSH as remote installation via SSH is fundamentally insecure. I'm not aware of any other distros that have this vulnerability.

**Comment by q66** [comment]
A solution that would satisfy everybody and be simple to implement is really a kernel command line parameter that would conditionally enable the service (all it takes is a single dracut hook)

**Comment by the-maldridge** [comment]
Hmm, perhaps I overestimated the knowledge of network level attacks when I wrote my original response.

If you do not trust the network you are installing on and doing the initial upgrade from, you can go ahead and just assume that your system has been compromised, ssh or not.  Void doesn't pin the public keys or even enforce HTTPS for the mirrors, so if your network is doing evil things, you aren't protected from it.  At some point you will have to trust the network you are installing to.

**Comment by noproto** [comment]
@the-maldridge I'm not sure who that was directed at, certainly not everyone else on this issue. Pointing out less secure parts of Void Linux does not clear the severity of this issue. Whereas your scenarios require a compromise of the network itself, this issue (40) only requires being on the network. Many networks do not filter traffic to the same subnet, so if I see someone (for example) installing Void Linux in a corporate red teaming environment I can connect to the WiFi and compromise their system for lateral movement -- without access to a router or backbone.

**Comment by Vaelatern** [comment]
noproto, I do not believe we have PermitRootLogin enabled in the live images.

I'm going to try it a little later, but I don't expect much.

**Comment by noproto** [comment]
@Vaelatern Tested twice on void-live-x86_64-20191109-xfce.iso

**Comment by hippi777** [comment]
hi folks! :)

(tldr below)

making an ssh accessible installer can have various valid reasons even if void inst intended to be a highly reliable server distro, like what if i have a private home server that i dont care if it breaks sometimes, as there wont be angry clients waiting for their service to be available again...

network must never be trusted, no matter if i trust in the corp or the ppl around me, their machines can be infected without their knowledge, and i wont investigate but let them live their life with their stuffs as they use them...

also, root or not root access is meaningless when anon has sudo :D

http mirrors arent mandatory, if someone wanna shoot their own feet, then who cares? the docs clearly states everything that is needed to be known, the defaults are fine.

tldr:
so my idea is to simply make the "voidlinux" pw modifiable, and the ssh optional for the iso builder stuffs, and in that case it can be trusted, except if there is something that i just forgot about

**Comment by q66** [comment]
I have tested this with current mklive and root login is definitely not enabled: https://imgur.com/rzB2Azf.png

not that it matters, since the environment has a pre-set root password (and anon has sudo)

That said, I still think this attack is fairly theoretical since it requires 1) knowing you're going to install Void on that machine 2) knowing you're going to install it at that specifc time 3) having hostile intention and you as a specific target 4) be either on the local network, or on a network that's not firewalled

**Comment by 0xdade** [comment]
I just grabbed `void-live-x86_64-20191109-xfce.iso`, booted into it, and PermitRootLogin is, in fact, enabled.

![looks enabled to me](https://i.gyazo.com/bed6faf1f1001f715a34e99ffbc853a6.png)

**Comment by hippi777** [comment]
>fairly theoretical
right, but why let it open without being necessary, and if its unavailable, then why running it unnecessary? (i dunno what about the anon user)

**Comment by hippi777** [comment]
btw there are things like portscanning, like default ssh port and dictionary attacks, and also, its not a good practice to say a lethal vulnerability is fine even if there isnt much chance... better to sleep well at the end of the day :D

**Comment by q66** [comment]
@0xdade that's a void-packages issue, not a mklive issue, currently fixed:

https://github.com/void-linux/void-packages/commit/5ce7496691e02dd264adc2f86c570c38a9976fc6
https://github.com/void-linux/void-packages/commit/ca84fd0a9aea0dd94dde6fd1f9a6a509cec5a789

but again, it does not matter since anon effectively has root anyway.

**Comment by Vaelatern** [comment]
sshd will be enabled by default. There may be a case to be made for not enabling it by default on the live ISOs that present graphical interfaces. If you want to make that case, please open another issue. Further discussion on this issue will result in the issue being locked.

**Comment by hippi777** [comment]
in that case, my last words to the topic:

actually im in the middle of making my own void installer, so i just wanted to sing it to others, that i wont take that role to open that issue, in case if that was addressed to me, and anyone would wait for me :D

(btw i think the discussion is still ontopic, but i dunno if reopening is an acceptable move here, or what ur workflow is; otherwise sorry for the additional noise! :) )

**Comment by camoz** [comment]
Hi, I want to politely ask if this issue can be reopened? AFAICS discussion on this has mostly been constructive.

@Vaelatern May I ask if there has been general consensus on the void team about enabling SSH by default?

What about q66's suggestion?:

> A solution that would satisfy everybody and be simple to implement is really a kernel command line parameter that would conditionally enable the service (all it takes is a single dracut hook)

To me that sounds easy to implement and easy for the end user.

Another idea to mitigate this issue: What's the point of the standard password `voidlinux` anyway, since it's publicly known? What about removing it? A local attacker targeting the installation process seems much less likely than an attacker on the network (which could obviously be trivially automated). And since `PermitEmptyPasswords no` is set by default in `sshd_config`, the live system would not be vulnerable by default anymore. All the user would have to do to enable remote login is setting a root password.

Btw, the live images might also be used by end users to rescue a system in production, not just to install a new system, so IMO it's worth getting it secure.

I also want to add that in my view, a possible attack is not "fairly theoretical" (as q66 thinks) but rather "fairly straightforward".

It surprises me that so many voidlinux members seem (or seemed?) to be okay with the current situation.

**Comment by Vaelatern** [comment]
`anon` password being random means you need a screen to log in, thus making it harder for a person to log in for headless installation. Not all of our gear has a KVM.

We're OK with an sshd in installation media. If you need an autoinstaller, we have one (and I've used it pretty heavily). Meanwhile more often I'm bitten by the installed machine not having sshd enabled, unless I manually enable it during the installation!

This approach works.

As for q66's points, they are all correct.  To list them again, an attack requires:

1. knowing you're going to install Void on that machine
2. knowing you're going to install it at that specific time
3. having hostile intention and you as a specific target
4. be either on the local network, or on a network that's not firewalled

If you need a non-general rescue tool, feel free to build in your SSH keys, and/or add users.. void-mklive exists, and you may customize your personal images however you like. For our users (ourselves) in the most likely environments, this works.

This issue will not be reopened. Further discussion will result in locking the issue. (yes, I said that last time, but traffic has been slow enough, and the point can be repeated).

**Comment by camoz** [comment]
I see. I accept that this will not be re-opened and sshd will remain enabled by default, but I will address some of the points of your last response.

> anon password being random means you need a screen to log in

I didn't suggest that. What I meant was removing the password altogether, see my last answer.

> If you need an autoinstaller, we have one (and I've used it pretty heavily). Meanwhile more often I'm bitten by the installed machine not having sshd enabled, unless I manually enable it during the installation!

I don't really understand how that has anything to do with this issue.

> As for q66's points, they are all correct.

I object:

> knowing you're going to install Void on that machine

No, since the attack can be trivially automated, as I have noted in my last answer. Have you never heard of malware randomly trying known standard passwords on SSH port(s) like `22`?

> knowing you're going to install it at that specific time

Not really, see last point.

> having hostile intention and you as a specific target

Yes, an attack requires hostile intentions...? As for the second part "you as a specific target" -> no, see previous points.

> be either on the local network, or on a network that's not firewalled

The only valid point. Still, IMO that is not enough to classify the attack as "theoretical".

**Comment by noproto** [comment]
You're correct @camoz, most of us on this earlier came to the same conclusion.

All the same, this is the beauty of Linux. If one distro doesn't prioritize security issues you can choose to use another or fork it. Void doesn't want to fix this so I chose to use a safer operating system.
